src/main/java/com/google/devtools/build/lib/query2/query/output/JSONProtoOutputFormatter.java

```diff
@@ -40,10 +42,13 @@ public OutputFormatterCallback<Target> createPostFactoStreamCallback(
       @Override
       public void processOutput(Iterable<Target> partialResult)
           throws IOException, InterruptedException {
+        List<String> jsonProtoBuffers = new ArrayList<>();
         for (Target target : partialResult) {
-          out.write(
-              jsonPrinter.print(toTargetProtoBuffer(target)).getBytes(StandardCharsets.UTF_8));
+          jsonProtoBuffers.add(jsonPrinter.print(toTargetProtoBuffer(target)));
         }
+        out.write("[".getBytes(StandardCharsets.UTF_8));
```

Comment:
As mentioned in https://github.com/bazelbuild/bazel/pull/18701#issuecomment-1607149702, instead of wrapping it in square brackets, follow the ndjson format and append a newline after the json message instead.